File: src/scala_module/BinaryTreeAsArray.scala
package scala_module

import java_module.{BinaryTree, DoWith}
import java_module.types.UserType

import java.util

class BinaryTreeAsArray extends BinaryTree {
  private var arr: Array[UserType] = new Array[UserType](0)
  private var size = 0 //размер массива
  private var level = 0

  def init(): Unit = {
    size = 1
    level = 0
    arr = new Array[UserType](size)
  }

  def this(array: Array[UserType]) {
    this()
    level = calculateLevelFromSize(array.length)
    size = calculateSizeFromLevel(level)
    arr = Array.copyOf(array, size)
  }

  def sizerLSS(n: Int): Int = {
    if (n >= size || arr(n) == null) return 0
    1 + sizerLSS(2 * n + 1) + sizerLSS(2 * n + 2)
  }

  def getByIndex(m: Int, n: Int): UserType = {
    if (m < 0 || m >= size || m >= sizerLSS(n)) return null
    val ll = sizerLSS(2 * n + 1)
    if (m < ll) return getByIndex(m, 2 * n + 1)
    val mid = m - ll
    val mid2 = mid-1
    if (mid2 == 0) return arr(n)
    getByIndex(mid, 2 * n + 2)
  }

  def insertByIndex(n: Int, element: UserType): Unit = {
    if (element == null) return
    if (n >= size) {
      level += 1
      size += calculateSizeOfLevel(level)
      arr = Array.copyOf(arr, size)
    }
    if (arr(n) == null) {
      arr(n) = element
      return
    }

    if (element.getTypeComparator.compare(element, arr(n)) > 0)
      insertByIndex(2 * n + 2, element) //право
    else insertByIndex(2 * n + 1, element) //влево
  }

  def getSize = size

  //балансировка
  def balance(array: util.ArrayList[UserType], a: Int, b: Int): Unit = {
    if (a > b) return
    val m = (a + b) / 2
    insertByIndex(0, array.get(m))
    balance(array, a, m - 1)
    balance(array, m + 1, b)
  }

  def sizer(n: Int, arrayList: util.ArrayList[UserType]): Int = { //подсчет потомков
    if (n >= size || arr(n) == null) return 0
    arrayList.add(arr(n))
    1 + sizer(2 * n + 1, arrayList) + sizer(2 * n + 2, arrayList)
  }

  def balance(): Unit = {
    val list = new util.ArrayList[UserType]
    val size1 = sizer(0, list)
    list.sort(list.get(0).getTypeComparator())
    init()
    balance(list, 0, size1 - 1)
  }

  //удаление
  def deleteByIndex(index: Int): Unit = {
    if (arr(index) == null) return
    delete(0, arr(index))
  }

  def delete(root: Int, element: UserType): UserType = {
    if (2 * root + 2 >= size) return null
    if (arr(root) == null) return arr(root)
    if (element.getTypeComparator.compare(element, arr(root)) < 0) { //если элемент < зн. узла = к левому потомку
      arr(2 * root + 1) = delete(2 * root + 1, element)
      return arr(root)
    }
    if (element.getTypeComparator().compare(element, arr(root)) > 0) { //если элемент > зн. узла = к правому потомку
      arr(2 * root + 2) = delete(2 * root + 2, element)
      return arr(root)
    }
    // если element найден
    if (arr(2 * root + 1) == null) {
      val temp = 2 * root + 2
      arr(root) = null
      return arr(temp)
    }
    if (arr(2 * root + 2) == null) {
      val temp = 2 * root + 1
      arr(root) = null
      return arr(temp)
    }
    arr(2 * root + 2) = deleteHelper(2 * root + 2, root) //если есть оба потомка
    arr(root)
  }

  def deleteHelper(root: Int, root0: Int): UserType = {
    if (2 * root + 2 < size) if (arr(2 * root + 1) != null) {
      arr(2 * root + 1) = deleteHelper(2 * root + 1, root0)
      return arr(root)
    }
    arr(root0) = arr(root)
    arr(root) = null
    if (2 * root + 2 < size) arr(2 * root + 2)
    else null
  }

  def show(): Unit = {
    var i = 0
    var cnt = 0
    var lvl = 0
    while (i < size) {
      if (i == cnt) {
        cnt += Math.pow(2, lvl).asInstanceOf[Int]
        lvl += 1
        System.out.println()
      }
      if (arr(i) == null) {
        System.out.print("N ")
      } else {
        System.out.print(arr(i).toString + " ")
      }
      i += 1
    }
  }

  override def toString = {
    var str = ""
    var i = 0
    var cnt = 0
    var lvl = 0
    while (i < size) {
      if (i == cnt) {
        cnt += Math.pow(2, lvl).asInstanceOf[Int]
        lvl += 1
        str += "\n"
      }
      if (arr(i) == null) {
        str += "N "
      }
      else {
        str += (arr(i).toString + " ")
        i += 1
      }
    }
    str
  }

  private def calculateLevelFromSize(sz: Int) = {
    var sum: Int = 0
    var level = -1
    while (sum < sz) {
      level += 1
      sum += Math.pow(2, level).asInstanceOf[Int]
    }
    level
  }

  private def calculateSizeFromLevel(level: Int) = {
    var sz = 0
    for (i <- 0 to level) {
      sz += Math.pow(2, i).asInstanceOf[Int]
    }
    sz
  }

  private def calculateSizeOfLevel(level: Int) = Math.pow(2, level).toInt

  def forEach(action: DoWith): Unit = {
    for (i <- 0 until arr.length) {
      var str: String = null
      if (arr(i) == null) str = "null "
      else str = arr(i).toString + " "
      action.doWith(str)
    }
  }
}

File: src/main/java/java_module/BinaryTreeAsArray.java
package java_module;

import java_module.types.UserType;

import java.util.ArrayList;


public class BinaryTreeAsArray implements BinaryTree {
    private ArrayList<UserType> arr;
    private int size;
    private int level;

    BinaryTreeAsArray() {
        init();
    }

    public void init() {
        size = 1;
        level = 0;
        arr = new ArrayList<>();
        for (int i = 0; i< size; i++) {
            arr.add(null);
        }
    }

    BinaryTreeAsArray(UserType[] array) {
        level = calculateLevelFromSize(array.length);
        size = calculateSizeFromLevel(level);
        for (int i=size/2; i < size; i++)
            arr.add(null);
    }

    int sizerLSS(int n) {
        if (n>=size || arr.get(n)==null) return 0;
        return 1 + sizerLSS(2*n+1)+sizerLSS(2*n+2);
    }

    public UserType getByIndex(int m, int n) {
        if (m < 0 || m >= size || m >= sizerLSS(n)) return null;
        int ll = sizerLSS(2 * n + 1);
        if (m < ll) return getByIndex(m, 2 * n + 1);
        m -= ll;
        if (m-- == 0) return arr.get(n);
        return getByIndex(m, 2 * n + 2);
    }

    public void insertByIndex(int n, UserType element) {
        if (element == null) return;

        if (n >= size) {
            level++;
            int oldSize = size;
            size += calculateSizeOfLevel(level);
            for (int i = oldSize+1; i<= size; i++)
                arr.add(null);
        }

        if (arr.get(n) == null) {
            arr.set(n, element);
            return;
        }

        if (element.getTypeComparator().compare(element, arr.get(n)) > 0)
            insertByIndex(2*n + 2, element); //право
        else
            insertByIndex(2*n + 1, element); //влево
    }

    int getSize() {return size;}

    //балансировка
    void balance(ArrayList<UserType> array, int a, int b) {
        if (a>b) return;
        int m = (a + b) / 2;
        insertByIndex(0, array.get(m));
        balance(array, a, m-1);
        balance(array, m+1, b);
    }

    int sizer(int n, ArrayList<UserType> arrayList) {//подсчет потомков
        if (n>=size || arr.get(n)==null) return 0;
        arrayList.add(arr.get(n));
        return 1 + sizer(2*n+1,arrayList)+sizer(2*n+2,arrayList);
    }

    int getSumPathLength(int n, int i) {
        if (n>=size || arr.get(n) == null) {
            return 0;
        }
        return i + getSumPathLength(2*n+1, i+1) + getSumPathLength(2*n+2, i+1);
    }

    int heigh(int n)
    {
        if (n>= size || arr.get(n) == null) {
            return 0;
        }
        else if (heigh(2*n+1) > heigh(2*n+2))
            return heigh(2*n+1) + 1;
        else
            return heigh(2*n+2) + 1;
    }

    public void balance() {
        ArrayList<UserType> list = new ArrayList<>();
        int  size1 = sizer(0,list);
        list.sort(list.get(0).getTypeComparator());
        init();
        balance(list, 0, size1-1);
    }

    //удаление
    public void deleteByIndex(int index) {
        if (arr.get(index) == null) return;
        delete(0,arr.get(index));
    }

    UserType delete(int root, UserType element) {
        if (2*root+2>=size) return null;
        if (arr.get(root) == null) return arr.get(root);


        if (element.getTypeComparator().compare(element, arr.get(root)) < 0) { //если элемент < зн. узла = к левому потомку
            arr.set(2*root + 1,delete(2*root + 1, element));
            return arr.get(root);
        }

        if (element.getTypeComparator().compare(element, arr.get(root)) > 0) {//если элемент > зн. узла = к правому потомку
            arr.set(2*root + 2,delete(2*root + 2, element));
            return arr.get(root);
        }

        if (arr.get(2*root+1) == null) {
            int temp = 2*root+2;
            arr.set(root, null);
            return arr.get(temp);
        }
        if (arr.get(2*root+2) == null) {
            int temp = 2*root+1;
            arr.set(root, null);
            return arr.get(temp);
        }

        arr.set(2*root+2, deleteHelper(2*root+2,root));
        return arr.get(root);
    }

    UserType deleteHelper(int root, int root0) {
        if (2*root+2<size) {
            if (arr.get(2*root+1) != null) {
                arr.set(2*root+1,deleteHelper(2*root+1, root0));
                return arr.get(root);
            }
        }

        arr.set(root0, arr.get(root));
        arr.set(root, null);
        if (2*root+2<size)  return arr.get(2*root + 2);
        else return null;
    }

    public void show() {
        for (int i = 0, cnt = 0, lvl = 0; i < size; i++) {
            if (i == cnt) {
                cnt += Math.pow(2, lvl);
                lvl++;
                System.out.println();
            }

            if (arr.get(i) == null) {
                System.out.print("N ");
                continue;
            }
            System.out.print(arr.get(i).toString() + " ");
        }
    }

    public String toString() {
        String str ="";
        for (int i = 0, cnt = 0, lvl = 0; i < size; i++) {
            if (i == cnt) {
                cnt += Math.pow(2, lvl);
                lvl++;
                str+="\n";
            }

            if (arr.get(i) == null) {
                str+="N ";
                continue;
            }
            str += (arr.get(i).toString() + " ");
        }
        return str;
    }

    private int calculateLevelFromSize(int sz) {
        int sum = 0;
        int level = -1;
        while (sum < sz) {
            level++;
            sum += Math.pow(2, level);
        }
        return level;
    }

    private int calculateSizeFromLevel(int level) {
        int sz = 0;
        for (int i = 0; i <= level; i++) {
            sz += Math.pow(2, i);
        }
        return sz;
    }

    private int calculateSizeOfLevel(int level) {
        return (int) Math.pow(2, level);
    }

    public void forEach(DoWith action) {
        for (int i=0; i<arr.size();i++) {
            String str;
            if (arr.get(i) == null) str = "null ";
            else str = arr.get(i).toString() + " ";
            action.doWith(str);
        }
    }

}
